checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallConsistencyAnalyzer.java

```diff
@@ -1787,197 +1958,247 @@ private boolean shouldTrackReturnType(MethodInvocationNode node) {
     // worklist if any of its resource aliases are still in scope in the successor block. If
     // none are, then the loop performs a consistency check for that Obligation.
     for (IPair<Block, @Nullable TypeMirror> successorAndExceptionType :
         getSuccessorsExceptIgnoredExceptions(currentBlock)) {
-      Block successor = successorAndExceptionType.first;
-      // If nonnull, currentBlock is an ExceptionBlock.
-      TypeMirror exceptionType = successorAndExceptionType.second;
-      // successorObligations eventually contains the Obligations to propagate to successor.
-      // The loop below mutates it.
-      Set<Obligation> successorObligations = new LinkedHashSet<>();
-      // A detailed reason to give in the case that the last resource alias of an Obligation
-      // goes out of scope without a called-methods type that satisfies the corresponding
-      // must-call obligation along the current control-flow edge. Computed here for
-      // efficiency; used in the loop over the Obligations, below.
-      String exitReasonForErrorMessage =
-          exceptionType == null
-              ?
-              // Technically the variable may be going out of scope before the method
-              // exit, but that doesn't seem to provide additional helpful
-              // information.
-              "regular method exit"
-              : "possible exceptional exit due to "
-                  + ((ExceptionBlock) currentBlock).getNode().getTree()
-                  + " with exception type "
-                  + exceptionType;
-      // Computed outside the Obligation loop for efficiency.
-      AccumulationStore regularStoreOfSuccessor = analysis.getInput(successor).getRegularStore();
-      for (Obligation obligation : obligations) {
-        // This boolean is true if there is no evidence that the Obligation does not go out
-        // of scope - that is, if there is definitely a resource alias that is in scope in
-        // the successor.
-        boolean obligationGoesOutOfScopeBeforeSuccessor = true;
-        for (ResourceAlias resourceAlias : obligation.resourceAliases) {
-          if (aliasInScopeInSuccessor(regularStoreOfSuccessor, resourceAlias)) {
-            obligationGoesOutOfScopeBeforeSuccessor = false;
-            break;
-          }
+
+      // A *mutable* set that eventually holds the set of dataflow facts to be propagated to
+      // successor blocks. The set is initialized to the current dataflow facts and updated by
+      // the methods invoked in the for loop below.
+      Set<Obligation> obligations = new LinkedHashSet<>(incomingObligations);
+
+      for (Node node : currentBlock.getNodes()) {
+        if (node instanceof AssignmentNode) {
+          updateObligationsForAssignment(obligations, cfg, (AssignmentNode) node);
+        } else if (node instanceof ReturnNode) {
+          updateObligationsForOwningReturn(obligations, cfg, (ReturnNode) node);
```

Comment:
We now redundantly compute these obligation updates once per successor block.  On the other hand, given that `obligations` is mutable, it's not obvious how to cache these results.  And, I don't even know if these computations are all that expensive.  Maybe we should add a comment here indicating we are at least aware of the redundancy and that optimization may be worthwhile in the future if we find this is a bottleneck?